### ISSUE TITLE
InjectMocks have naive approach in choosing the multiple constructors with same size 

### ISSUE THREAD

**Issue by xinyi9** [mockito/mockito#1935]
Hey Mockito Team,
 
Apologies if similar problem has been reported somewhere. 

In our test, we have 
`
@RunWith(MockitoJUnitRunner.Silent.class)
public class TestCachingBarDao { 
    @Mock
    BarIdCache mockBarIdCache;
    @Mock
    BarDaoInternal mockDelegateDao;
    @InjectMocks
    private CachingFooDao cachingBarDao;
 
    //some other Mock...
    @Mock
    Foo foo;

    @Mock
    FooById fooById; 
 
 
    @Before
    public void setUp() throws Exception {
        ....
    }
 
    @Test
    ...
 
     
    @Test
    ...
}
`

Whereby the testing class is

`
public class CachingBarDao extends DelegatingBarDao {
    private final BarIdCache barIdCache;`
 
    public CachingBarDao(BarDaoInternal delegateDao, BarIdCache barIdCache) {
        super(delegateDao);
        this.barIdCache = barIdCache;
    }
 
    public CachingBarDao(BarDaoInternal delegateDao, VCacheFactory cacheFactory) {
        this(delegateDao, new BarIdCache(Lazy.supplier(() -> createCache(cacheFactory))));
    }
` 
}
` 

Recently we've discovered the tests have been failing intermittently, after some head scratching investigation, we found out it arose from the approach that InjectMocks chose the biggest constructor for Injection. 
After debugging into the problem, I found the implementation of the biggestConstructor is pretty naive.  https://github.com/mockito/mockito/blob/release/3.x/src/main/java/org/mockito/internal/util/reflection/FieldInitializer.java 
Getting declared constructors, sort constructors by parameters number and get the biggest constructor, whereby in our case, the mockableParam count are same for both constructors. It ties up to the order of what returned from  clazz.getDeclaredConstructors(), and it has caused us problem in different CI pipeline. 

Wondering could your team 
1. improve the biggestConstructor approach to take constructor signature into account? 
2. Update on the Javadoc on the potential caveat on same size constructors using InjectMocks 

Thank you and look forward to hear from you.

**Comment by martinnemec3** [comment]
I agree with this proposal. I am wondering what is the reason for choosing the constructor this way. My expectation would be to use the information about @Mock-ed fields to choose the constructor appropriately and only if there is no one that would match, use the biggestConstructor() or similar procedure as a fallback.

**Comment by atl-mk** [comment]
Just ran into this today
